### PR TITLE
feat(chat-api): add attached dial prompt skill to be shared with the quick app (#8529)

### LINKED ISSUE #8529: [Chat] [Sharing] Attached prompt not shared with the application — "Access denied to the prompt"
### DIAL chat version
1.0.0

### How to reproduce
1. Create/open an application with a prompt attached to it.
2. Share the application with another user.
3. As the recipient, open the shared application and try to use it (which relies on the attached prompt).

### Actual result
The recipient gets "Access denied to the prompt" — the prompt attached to the application is not shared along with the application itself. Reproduces every time.

### Expected result
Sharing an application should also grant access to the prompt attached to it, so the recipient can use the application without an access error.

### Confidential information
- [x] I confirm that do not share any confidential information

### PR DESCRIPTION
**Pull request by Yaroslav-Hapon**
## Summary

Fixes #8529 — sharing a quick app (an application built on a quick-app schema) shared only the application resource itself. The prompt attached to it as a `dial-prompt` skill (`application_properties.skills[]`) is a separate DIAL Core prompt resource that received no share grant, so the recipient hit "Access denied to the prompt."

`ShareService.createShareLink` already resolved related resources for conversations (file attachments) but did no lookup for applications. The `conversation-share` spec explicitly mandated "no related-resource lookup" for every non-conversation kind, which left quick apps' attached prompts unshared.

## Changes

**`apps/chat-api/src/share/chat-api` only — no frontend change, no new endpoint, no DTO change.**

- `ShareService.getRelatedResourceUrls` now dispatches by resource kind:
  - `conversations/` → existing file-attachment walk (extracted unchanged into `getConversationRelatedResourceUrls`)
  - `applications/` → new `getApplicationRelatedResourceUrls`: parses `applications/{bucket}/{path}`, calls `getCustomApplication`, collects `application_properties.skills[]` entries where `type === 'dial-prompt'`, and appends their `url` to the share request with the same permissions as the app.
  - everything else → `[]`, proxied directly (unchanged).
- Same cross-bucket rule as conversations: a referenced prompt in another user's private bucket is silently dropped (DIAL Core rejects mixed-owner share requests); a `public`-bucket prompt is kept.
- Related prompt urls pass through `toShareResourceUrl` (encoding) — a prompt stored raw in `skills[].url` (e.g. with a space) is encoded before hitting DIAL Core, which rejects unencoded paths with 400.
- `#`-fragment suffixes stripped from skill urls, matching the conversation attachment collector.
- `collectApplicationPromptResourceUrls` is a defensive pure helper (opaque `application_properties`, no import of the editor's `QuickApp2Config` type); covered by direct unit tests for its edge cases.

## Test plan

- [x] `share.service.spec.ts` — integration tests (single prompt shared, dedup, cross-bucket drop, public-bucket keep, no skills, no application_properties, lookup reject/error/no-data) + direct unit tests for `collectApplicationPromptResourceUrls`.
- [x] `nx lint chat-api` clean.
- [x] `nx build chat-api` clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: apps/chat-api/src/share/share.service.ts [hunk 1]
```diff
@@ -296,6 +377,63 @@ export class ShareService {
...
+      );
+    }
+    if (result.data == null) {
+      this.logger.warn(
+        `DIAL Core returned an empty application for resourceUrl=${resourceUrl}; sharing without related prompts`,
+      );
+      return [];
+    }
+
+    return collectApplicationPromptResourceUrls(result.data).filter((url) => {
```
[the remark this answers is not in the document]
Comment (Yaroslav-Hapon) [reply apps/chat-api/src/share/share.service.ts:443]: fixed

File: apps/chat-api/src/share/share.service.ts [hunk 2]
```diff
@@ -188,6 +209,48 @@ const collectConversationResourceUrls = (conversation: unknown): string[] => {
...
+    if (typeof entry.url !== 'string') continue;
+
+    /*
+     * Strip a `#`-fragment the same way `collectAttachmentResourceUrls` does
+     * for file attachments: a fragment identifies a view within a resource
+     * (e.g. `prompts/.../my-prompt#variables`), not a distinct resource, and
+     * leaving it on would either be rejected by DIAL Core or grant against a
+     * non-existent path.
+     */
+    const resourceUrl = entry.url.split('#', 1)[0];
```
[the remark this answers is not in the document]
Comment (Yaroslav-Hapon) [reply apps/chat-api/src/share/share.service.ts:241]: fixed

File: apps/chat-api/src/share/share.service.ts [hunk 3]
```diff
@@ -188,6 +209,59 @@ const collectConversationResourceUrls = (conversation: unknown): string[] => {
...
+    /*
+     * Path-traversal guard: `encodeURIComponent` does not encode `.`, so a
+     * `..` segment survives `encodeDialResourcePath` unchanged and could reach
+     * DIAL Core's share endpoint. DIAL Core enforces resource ownership itself,
+     * so this is not exploitable for cross-user escalation, but a `..` segment
+     * in a prompt resource url is never legitimate — dropping it here is
+     * defense-in-depth, consistent with the `@IsValidFilePath()` /
+     * `hasTraversalSegment` guards applied to user-supplied itemId values.
+     */
+    if (resourceUrl.split('/').includes('..')) continue;
```
[the remark this answers is not in the document]
Comment (Yaroslav-Hapon) [reply apps/chat-api/src/share/share.service.ts]: fixed

File: apps/chat-api/src/share/share.service.ts [hunk 4]
```diff
@@ -188,6 +209,59 @@ const collectConversationResourceUrls = (conversation: unknown): string[] => {
...
+
+    resourceUrls.add(resourceUrl);
```
[the remark this answers is not in the document]
Comment (Yaroslav-Hapon) [reply apps/chat-api/src/share/share.service.ts]: fixed

File: apps/chat-api/src/share/share.service.ts [hunk 5]
```diff
@@ -249,8 +323,26 @@ export class ShareService {
...
-    if (!resourceUrl.startsWith(CONVERSATION_RESOURCE_PREFIX)) return [];
+    if (resourceUrl.startsWith(CONVERSATION_RESOURCE_PREFIX)) {
+      return this.getConversationRelatedResourceUrls(
+        accessToken,
+        sessionBucket,
+        resourceUrl,
+      );
+    }
+
+    if (resourceUrl.startsWith(APPLICATION_RESOURCE_PREFIX)) {
```
[the remark this answers is not in the document]
Comment (Yaroslav-Hapon) [reply apps/chat-api/src/share/share.service.ts:329]: fixed

File: apps/chat-api/src/share/share.service.ts [hunk 6]
```diff
@@ -188,6 +209,69 @@ const collectConversationResourceUrls = (conversation: unknown): string[] => {
...
+     * Normalize through `toShareResourceUrl` before deduping so two skill
+     * entries that differ only in percent-encoding (`prompts/bucket/My prompt`
+     * and `prompts/bucket/My%20prompt`) collapse to one Set entry instead of
+     * producing a duplicate in `shareResource.resources` after the caller
+     * encodes them. `encodeDialResourcePath` is idempotent (it decodes each
+     * segment first), so an already-encoded url is not double-encoded. This
+     * also means the bucket filter below reads the same encoded form DIAL Core
+     * receives, not the raw unencoded one.
+     */
+    resourceUrls.add(toShareResourceUrl(resourceUrl));
```
[the remark this answers is not in the document]
Comment (Yaroslav-Hapon) [reply apps/chat-api/src/share/share.service.ts]: fixed

File: apps/chat-api/src/share/share.service.ts [hunk 7]
```diff
@@ -188,6 +209,69 @@ const collectConversationResourceUrls = (conversation: unknown): string[] => {
...
+ * Collects unique prompt resource urls referenced by a quick app's
+ * `application_properties.skills[]`. Only `type: 'dial-prompt'` entries carry
+ * a `url` pointing at a separate DIAL prompt resource (`prompts/{bucket}/...`)
+ * that has to be granted alongside the app for the recipient to use it — the
+ * `custom` system prompt and every other skill kind hold their content inline
+ * and share nothing. `application_properties` is opaque to this repo (authored
+ * by the embedded Quick Apps editor), so the walk is defensive: a missing or
+ * malformed `skills` array yields no urls rather than throwing.
+ */
+export const collectApplicationPromptResourceUrls = (
```
[the remark this answers is not in the document]
Comment (Yaroslav-Hapon) [reply apps/chat-api/src/share/share.service.ts:224]: fixed

File: apps/chat-api/src/share/share.service.ts [hunk 8]
```diff
@@ -296,6 +377,79 @@ export class ShareService {
...
+       * `toShareResourceUrl`, so the bucket segment read here is the encoded
+       * form (e.g. `My%20Bucket`). `resource.bucket` is the raw bucket
+       * segment from the application's itemId, which may itself be encoded or
+       * unencoded depending on the caller. Decode both before comparing so a
+       * percent-encoded bucket name (S3-style) matches regardless of which
+       * side carries the encoding.
+       */
+      const promptBucket = safeDecodeURIComponent(getResourceBucket(url));
+      const appBucket = safeDecodeURIComponent(resource.bucket);
+      return promptBucket === appBucket || promptBucket === PUBLIC_BUCKET;
```
[the remark this answers is not in the document]
Comment (Yaroslav-Hapon) [reply apps/chat-api/src/share/share.service.ts:455]: fixed

File: openspec/changes/archive/2026-09-07-share-attached-quick-app-prompt/proposal.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+- Any new endpoint or DTO.
+
+## Acceptance criteria
+
+- Sharing a quick app with one `dial-prompt` skill shares the app and the prompt
+  with the same permissions.
+- Multiple `dial-prompt` skills are shared once each (deduped, first-seen order);
+  `type: 'custom'` and other skill kinds contribute nothing.
+- A referenced prompt in another user's private bucket is omitted; the share
+  still succeeds. A prompt in the `public` bucket is kept.
```
[the remark this answers is not in the document]
Comment (Yaroslav-Hapon) [reply openspec/changes/archive/2026-09-07-share-attached-quick-app-prompt/proposal.md:56]: fixed

File: apps/chat-api/src/share/share.service.ts [hunk 10]
```diff
@@ -188,6 +209,48 @@ const collectConversationResourceUrls = (conversation: unknown): string[] => {
...
+    if (!isRecord(entry)) continue;
+    if (entry.type !== DIAL_PROMPT_SKILL_TYPE) continue;
+    if (typeof entry.url !== 'string') continue;
+
+    // Strip `#`-fragments — a fragment identifies a view, not a distinct resource.
+    const resourceUrl = entry.url.split('#', 1)[0];
+    if (!isPromptResourceUrl(resourceUrl)) continue;
+
+    // Drop `..` traversal segments — `encodeURIComponent` doesn't encode `.`.
+    if (resourceUrl.split('/').includes('..')) continue;
```
[the remark this answers is not in the document]
Comment (Yaroslav-Hapon) [reply apps/chat-api/src/share/share.service.ts]: fixed

File: apps/chat-api/src/share/share.service.ts [hunk 11]
```diff
@@ -188,6 +209,51 @@ const collectConversationResourceUrls = (conversation: unknown): string[] => {
...
+    // Strip `#`-fragments — a fragment identifies a view, not a distinct resource.
+    const resourceUrl = entry.url.split('#', 1)[0];
+    if (!isPromptResourceUrl(resourceUrl)) continue;
+
+    // Normalize before dedup so encoded/decoded variants collapse to one entry.
+    const normalizedUrl = toShareResourceUrl(resourceUrl);
+
+    // Drop `..` traversal segments — check after normalization so a percent-
+    // encoded `%2E%2E` (which decodes to `..`) cannot bypass the guard.
+    if (normalizedUrl.split('/').includes('..')) continue;
```
[the remark this answers is not in the document]
Comment (Yaroslav-Hapon) [reply apps/chat-api/src/share/share.service.ts]: fixed

File: apps/chat-api/src/share/share.service.ts [hunk 12]
```diff
@@ -188,6 +209,54 @@ const collectConversationResourceUrls = (conversation: unknown): string[] => {
...
+  const resourceUrls = new Set<string>();
+  for (const entry of skills as DialPromptSkillEntry[]) {
+    if (!isRecord(entry)) continue;
+    if (entry.type !== DIAL_PROMPT_SKILL_TYPE) continue;
+    if (typeof entry.url !== 'string') continue;
+
+    // Strip `#`-fragments — a fragment identifies a view, not a distinct resource.
+    const resourceUrl = entry.url.split('#', 1)[0];
+    if (!isPromptResourceUrl(resourceUrl)) continue;
+
```
[the remark this answers is not in the document]
Comment (Yaroslav-Hapon) [reply apps/chat-api/src/share/share.service.ts:243]: No change needed — this is a redundant restatement of the fix we already applied. The guard is correct and covered by three traversal tests (literal .., %2E%2E, %2F-hidden).

File: apps/chat-api/src/share/share.service.ts [hunk 13]
```diff
@@ -296,6 +383,79 @@ export class ShareService {
...
+       * `toShareResourceUrl`, so the bucket segment read here is the encoded
+       * form (e.g. `My%20Bucket`). `resource.bucket` is the raw bucket
+       * segment from the application's itemId, which may itself be encoded or
+       * unencoded depending on the caller. Decode both before comparing so a
+       * percent-encoded bucket name (S3-style) matches regardless of which
+       * side carries the encoding.
+       */
+      const promptBucket = safeDecodeURIComponent(getResourceBucket(url));
+      const appBucket = safeDecodeURIComponent(resource.bucket);
+      return promptBucket === appBucket || promptBucket === PUBLIC_BUCKET;
```
[the remark this answers is not in the document]
Comment (Yaroslav-Hapon) [reply apps/chat-api/src/share/share.service.ts:455]: No change needed — the comparison is already correct. The reviewer's suggested safeDecodeURIComponent(PUBLIC_BUCKET) is a no-op ('public' → 'public').

File: openspec/specs/conversation-share/spec.md [hunk 14]
```diff
@@ -57,16 +57,57 @@ The backend `POST /api/v1/share` (`apps/chat-api/src/share/share.controller.ts`)
...
+#### Scenario: Cross-bucket private prompt is dropped
+
+- **GIVEN** a quick app references `prompts/other-user-bucket/theirs` via a `dial-prompt` skill
+- **WHEN** a share link is created for `applications/owner-bucket/my-app__1.0`
+- **THEN** `prompts/other-user-bucket/theirs` is omitted from `shareResource.resources` and the share still succeeds
+
+#### Scenario: Public-bucket prompt is kept
+
+- **GIVEN** a quick app references `prompts/public/shared` via a `dial-prompt` skill
+- **WHEN** a share link is created for `applications/owner-bucket/my-app__1.0`
```
[the remark this answers is not in the document]
Comment (Yaroslav-Hapon) [reply openspec/specs/conversation-share/spec.md:103]: The spec's "proxies directly, no lookup" wording is accurate as written. No change needed.

**Review by valerydluski — APPROVED** [review]
